tcljava/doc/TclInteger.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclInteger.3,v 1.1 1998/10/14 21:09:08 cvsadmin Exp $
'\" 
.so man.macros
.TH TclInteger 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclInteger -- This class is used to create, read, and modify integer
objects from Java code.

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR(int \fIintValue\fR)
.sp
static int \fBget\fR(Interp \fIinterp\fR, TclObject \fItobj\fR) throws TclException
.sp
static int \fBgetForIndex\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, int \fIendValue\fR) throws TclException
.sp
static void \fBset\fR(TclObject \fItobj\fR, int \fIintValue\fR)

.SH ARGUMENTS
.AS TclObject "intValue" .
.AP int intValue
New value for internal integer representation.
.AP Interp interp
If an error occurs while converting an object to be an integer object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
TclObject to manipulate.  If the internal representation of
\fItobj\fR is not TclInteger, the method converts it to TclInteger.
.AP int endValue
Value returned if \fItobj\fR is a String object whose value is "end".
.BE

.SH DESCRIPTION
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new TclObject whose internal
representation is the integer \fIintValue\fR.
.TP
\fBget\fR
The \fBget\fR method returns an integer value from the \fItobj\fR.  If an error
occurs while converting \fItobj\fR to be an integer object, an error
message is left in the \fIinterp\fR's result object unless
\fIinterp\fR is NULL.
.TP
\fBgetForIndex\fR
The \fBgetForIndex\fR method is intended to simplify operations that expect
either an integer value or the string "end".  If \fItobj\fR contains
the string "end", \fIendValue\fR is returned. Otherwise, the same
operations as TclInteger.get are performed.
.TP
\fBset\fR
The \fBset\fR method sets \fItobj\fR to have the integer value of \fIintValue\fR.
When this method returns, the internal representation of \fItobj\fR
is TclInteger.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewIntObj, Tcl_GetIntFromObj, TclGetIntForIndex, Tcl_SetIntObj

.SH "SEE ALSO"
InternalRep, TclObject, TclList, TclIndex, TclString, TclDouble, TclBoolean, ReflectObject

.SH KEYWORDS
internal representation, object, object type, integer


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/TclInteger.java
/*
 * TclInteger.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclInteger.java,v 1.16 2006/06/09 20:13:39 mdejong Exp $
 *
 */

package tcl.lang;

/**
 * This class implements the integer object type in Tcl.
 */

public class TclInteger implements InternalRep {

    // The int value for a TclInteger type is stored
    // in the TclObject instance. The TclObject API
    // requires that every TclObject have a non-null
    // internal rep, so this dummy value is used
    // for every TclObject with an int value.
    // The dummy value also maintains compatibility with
    // old code that might check for an internal
    // rep via:
    //
    // if (tobj.getInternalRep() instanceof TclInteger) {...}

    final static TclInteger dummy = new TclInteger();

    // Extra debug checking
    private static final boolean validate = false;

    private TclInteger() {}

    /**
     * Should never be invoked.
     */

    public InternalRep duplicate() {
	throw new TclRuntimeError("TclInteger.duplicate() should not be invoked");
    }

    /**
     * Implement this no-op for the InternalRep interface.
     */

    public void dispose() {}

    /**
     * Should never be invoked.
     */
    public String toString() {
	throw new TclRuntimeError("TclInteger.toString() should not be invoked");
    }

    /**
     * Tcl_NewIntObj -> TclInteger.newInstance
     *
     * Creates a new instance of a TclObject with a TclInteger internal
     * representation.
     *
     * @param b initial value of the integer object.
     * @return the TclObject with the given integer value.
     */

    public static TclObject newInstance(int i) {
	return new TclObject(i);
    }

    /**
     * SetIntFromAny -> TclInteger.setIntegerFromAny
     *
     * Called to convert the other object's internal rep to this type.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to convert to use the
     *     representation provided by this class.
     */

    private static void setIntegerFromAny(Interp interp, TclObject tobj)
	    throws TclException
    {
	// Note that this method is never invoked when the
	// object is already an integer type. This method
	// does not check for a TclBoolean internal rep
	// since it would typically only be used only for
	// the "true" and "false" string values. An
	// int value like "1" or "0" that can be a boolean
	// value will not be converted to TclBoolean.
	//
	// This method also does not check for a TclDouble
	// internal rep since a double like "1.0" can't
	// be converted to an integer. An octal like
	// "040" could be parsed as both a double and
	// an integer, but the TclDouble module should
	// not allow conversion to TclDouble in that case.

	int ivalue = Util.getInt(interp, tobj.toString());
	tobj.setInternalRep(dummy);
	tobj.ivalue = ivalue;

	if (TclObject.saveObjRecords) {
	    String key = "TclString -> TclInteger";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = new Integer(1);
	    } else {
	        num = new Integer(num.intValue() + 1);
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Tcl_GetIntFromObj -> TclInteger.get
     *
     * Returns the integer value of the object.
     *
     * @param interp current interpreter.
     * @param tobj the object to operate on.
     * @return the integer value of the object.
     */

    public static int get(
	final Interp interp,
	final TclObject tobj)
	    throws TclException
    {
	if (!tobj.isIntegerType()) {
	    setIntegerFromAny(interp, tobj);
	}
	return tobj.ivalue;
    }

    /**
     * Changes the integer value of the object.
     *
     * @param interp current interpreter.
     * @param tobj the object to operate on.
     * @param i the new integer value.
     */
    public static void set(TclObject tobj, int i) {
	if (!tobj.isIntegerType()) {
	    // Change the internal rep if not an integer.
	    // Note that this method does not reparse
	    // an int value from the string rep.
	    tobj.setInternalRep(dummy);
	}
	tobj.invalidateStringRep();
	tobj.ivalue = i;
    }

    /**
     * Increments the integer value of the object by the given
     * amount. One could implement this same operation by
     * calling get() and then set(), this method provides an
     * optimized implementation. This method is not public
     * since it will only be invoked by the incr command.
     *
     * @param interp current interpreter.
     * @param tobj the object to operate on.
     * @param incrAmount amount to increment
     */
    static void incr(
        final Interp interp,
        final TclObject tobj,
        final int incrAmount)
            throws TclException
    {
	if (!tobj.isIntegerType()) {
	    setIntegerFromAny(interp, tobj);
	}
	tobj.invalidateStringRep();
	tobj.ivalue += incrAmount;
    }

    /**
     * This special helper method is used only by
     * the Expression module. This method will
     * change the internal rep to a TclInteger with
     * the passed in int value. This method does
     * not invalidate the string rep since the
     * object's value is not being changed.
     *
     * @param tobj the object to operate on.
     * @param i the new int value.
     */

    static void exprSetInternalRep(
	final TclObject tobj,
	final int i)
    {
	if (validate) {

	    // Double check that the internal rep is not
	    // already of type TclInteger.

	    if (tobj.isIntegerType()) {
	        throw new TclRuntimeError("exprSetInternalRep() called with object" +
	            " that is already of type TclInteger");
	    }

	    // Double check that the new int value and the
	    // string rep would parse to the same integer.

	    int i2;
            try {
                i2 = Util.getInt(null, tobj.toString());
            } catch (TclException te) {
	        throw new TclRuntimeError("exprSetInternalRep() called with int" +
	            " value that could not be parsed from the string");
            }
	    if (i != i2) {
	        throw new TclRuntimeError("exprSetInternalRep() called with int value " +
	            i + " that does not match parsed int value " + i2 +
                    ", parsed from str \"" +
	            tobj.toString() + "\"");
	    }
	}

	tobj.setInternalRep(dummy);
	tobj.ivalue = i;
    }

}

